enigma-swing/src/main/java/cuchaz/enigma/gui/Gui.java

```diff
@@ -482,32 +485,47 @@ public void redraw() {
 		frame.repaint();
 	}
 
-	public void onRenameFromClassTree(ValidationContext vc, Object prevData, Object data, DefaultMutableTreeNode node) {
-		if (data instanceof String) {
-			// package rename
-			for (int i = 0; i < node.getChildCount(); i++) {
-				DefaultMutableTreeNode childNode = (DefaultMutableTreeNode) node.getChildAt(i);
-				ClassEntry prevDataChild = (ClassEntry) childNode.getUserObject();
-				ClassEntry dataChild = new ClassEntry(data + "/" + prevDataChild.getSimpleName());
+	public void onRenameFromClassTree(ValidationContext vc, String targetName, DefaultMutableTreeNode node) {
+		List<EntryChange<ClassEntry>> task = new ArrayList<>();
+		onRenameFromClassTree(targetName, node, task);
+		this.controller.applyChanges(vc, task);
```

Comment:
```suggestion
        task.forEach(c -> this.controller.validateChange(vc, task));
        
        if (!vc.canProceed()) {
            return;
        }
        
		task.forEach(c -> this.controller.applyChange(vc, task));
```

This is better because now all the changes will be validated and it won't bail out at the first bad one (the iterator in applyChanges you added doesn't validate the rest after one fails). This is how this should be used, validate everything first, then actually apply, and they are deliberately separate steps for batch operations like this.

You don't need to change anything about GuiController.

Reply:
Should be fixed.